includes/models/model.llms.student.php

```diff
@@ -1841,4 +1909,162 @@ private function update_completion_status( $status, $object_id, $object_type, $t
+	 * @param string $object_type Type, 'Lesson'.
+	 * @return bool
+	 */
+	public function is_favorite( $object_id, $object_type = 'lesson' ) {
+
+		$query = new LLMS_Query_User_Postmeta(
+			array(
+				'types'                 => 'favorites',
+				'include_post_children' => false,
+				'user_id'               => $this->get_id(),
+				'post_id'               => $object_id,
+				'per_page'              => 1,
+			)
+		);
+
+		$ret = $query->has_results();
+
+		/**
+		 * Filter object favorite boolean value prior to returning.
+		 *
+		 * The dynamic portion of this filter, `{$object_type}`, refers to the Lesson.
+		 *
+		 * @since [version]
+		 *
+		 * @param array|false  $ret         Array of favorite data or `false` if no favorite is found.
+		 * @param int          $object_id   WP Post ID of the Lesson.
+		 * @param string       $object_type Type, 'Lesson'.
+		 * @param LLMS_Student $instance    The Student Instance
+		 */
+		return apply_filters( 'llms_is_' . $object_type . '_favorite', $ret, $object_id, $object_type, $this );
+
+	}
+
+	/**
+	 * Mark a lesson favorite for the given user.
+	 *
+	 * @since [version]
+	 *
+	 * @see llms_mark_favorite() calls this function without having to instantiate the LLMS_Student class first.
+	 *
+	 * @param int    $object_id   WP Post ID of the Lesson.
+	 * @param string $object_type Type, 'Lesson'.
+	 * @return bool
+	 */
+	public function mark_favorite( $object_id, $object_type ) {
+
+		// Short circuit if it's already favorited.
+		if ( $this->is_favorite( $object_id, $object_type ) ) {
+			return true;
+		}
+
+		return $this->update_favorite_status( 'favorite', $object_id, $object_type );
+
+	}
+
+	/**
+	 * Mark a lesson unfavorite for the given user.
+	 *
+	 * @since [version]
+	 *
+	 * @see llms_mark_unfavorite() calls this function without having to instantiate the LLMS_Student class first.
+	 *
+	 * @param int    $object_id   WP Post ID of the Lesson.
+	 * @param string $object_type Type, 'Lesson'.
+	 * @return bool
+	 */
+	public function mark_unfavorite( $object_id, $object_type ) {
+
+		// Short circuit if it's not favorited.
+		if ( ! $this->is_favorite( $object_id, $object_type ) ) {
+			return true;
+		}
+
+		return $this->update_favorite_status( 'unfavorite', $object_id, $object_type );
+
+	}
+
+	/**
+	 * Triggers actions for favorite/unfavorite.
+	 *
+	 * Update the favorite status of a lesson for the current student.
+	 * Inserts / updates necessary user postmeta data.
+	 *
+	 * @since [version] Use filterable functions to determine if the object can be marked favorite.
+	 *
+	 * @param string $status      New status to update to, either "favorite" or "unfavorite".
+	 * @param int    $object_id   WP Post ID of the Lesson.
+	 * @param string $object_type Type, 'Lesson'.
+	 * @param string $trigger     String describing the reason for the status change.
+	 * @return bool
+	 */
+	private function update_favorite_status( $status, $object_id, $object_type, $trigger = 'unspecified' ) {
```

Comment:
do you need the trigger parameter here? what are you envisioning? How are you going to use it?

Reply:
I thought we can use provide an API endpoint to mark a lesson as favorite/unfavorite and save the "API" string as a trigger.

Reply:
I see.
I understand you based this on other stuff like enrollments and completions, in that case tracking the source (trigger) has a meaning, because different actors can be responsible for that, e.g. a lesson can be completed because of clicking on mark complete, or because the quiz/assignment has been completed... stuff like that.
For the enrollment it's more like was the user enrolled because of an admin enrolled the user manually, or because they bought the course, or because they bought a membership which auto-enrolled into the course - the latter is important because it will allow the unenroll when someone is unenrolled from the membership.

In this case I see no real value on tracking the trigger, it only adds complexity and stored data.
Actually you're not storing the trigger at all, since your `insert_favorite_postmeta` accepts only one parameter, and in any case has no logic to handle the trigger.
So let's get rid of the trigger :D

Reply:
Got it. I didn't use the trigger. Shit🤦‍♂️
I've removed it now.